### ISSUE TITLE
The Extension is Cutting off prompts on generation

### ISSUE THREAD

**Issue by Bonedragon1300** [adieyal/sd-dynamic-prompts#144]
**Describe the bug**
when dynamic prompts is on, it just cuts off after a few words, it was working just fine last night, but I updated it and now it doesn't work

**To Reproduce**
just turning on dynamic prompts, no other checkbox is on, not even restore faces and highres fix

**Expected behaviour**
that the prompt would be generated in full

**Additional context**
as an example, here is a the prompt attached to the output picture with dynamic prompts on

![Capture](https://user-images.githubusercontent.com/15267401/210185285-1032d75e-7f96-4fc8-a4a1-19571b2422bc.PNG)

((best quality)), ((highly detailed)), masterpiece, ((official art)), (detailed eyes, deep eyes), (1girl
Negative prompt: bad_prompt_version2, low quality, lowres, (bad anatomy, bad hands

and the same prompt once I turn it off

![Capture](https://user-images.githubusercontent.com/15267401/210185309-15e912fc-a235-429c-8794-1c2dee6ec31f.PNG)

((best quality)), ((highly detailed)), masterpiece, ((official art)), (detailed eyes, deep eyes), (1girl:1.33, solo), (dragon girl), (red scales), (dragon_wings), large wings, 1 dragon_tail, (((horns)), dragon_horns), large horns, (abs), (gold bracelets and anklets), armor, large breasts
Negative prompt: bad_prompt_version2, low quality, lowres, (bad anatomy, bad hands:1.1,bad proportions), worst quality, 3d, watermark, (bodysuit), missing fingers, extra digit, fewer digits, malformed hands, error, jpeg artifacts, cropped, no thumb, extra fingers, mutated hands, (deformation), deformed, (mutilated), text, signature, black and white, logo, (asymmetrical breasts), (cross eyed), ((bad eyes)), (disfigured eyes), (simple background), speech bubble, multiple panels, (head out of frame:1.2), (bird_head), (furry), (blob), malformed heels, asymmetrical heels,

**Comment by floorcat** [comment]
yeah, cuts prompt at first `:` for me as well

**Comment by adieyal** [comment]
Could you please test 1.5.14 - this should be fixed now.

**Comment by floorcat** [comment]
> Could you please test 1.5.14 - this should be fixed now.

yes, this is fixed

**Comment by adieyal** [comment]
@Bonedragon1300 could you confirm so that I can close this issue?

**Comment by tusharbhutt** [comment]
Heh... not for me

**Comment by Bonedragon1300** [comment]
made it worse for me actually, now it's cutting off at the whole word its attached to instead of at the :

**Comment by tusharbhutt** [comment]
I get a cutoff right at the colon

![image](https://user-images.githubusercontent.com/26264462/210186906-1c3f075a-ee68-41d6-be9a-18dd84cf0d41.png)

**Comment by adieyal** [comment]
> made it worse for me actually, now it's cutting off at the whole word its attached to instead of at the :

In your example, you use `(1girl:1.33, solo)` I don't think that's valid - the `:1.33` should be at the end shouldn't it?

**Comment by adieyal** [comment]
> I get a cutoff right at the colon
> 
> ![image](https://user-images.githubusercontent.com/26264462/210186906-1c3f075a-ee68-41d6-be9a-18dd84cf0d41.png)

Can you share your exact prompt?

**Comment by Bonedragon1300** [comment]
> > made it worse for me actually, now it's cutting off at the whole word its attached to instead of at the :
> 
> In your example, you use `(1girl:1.33, solo)` I don't think that's valid - the `:1.33` should be at the end shouldn't it?
 
it's always worked before, and besides, the :1.33 is for 1girl, not both

**Comment by tusharbhutt** [comment]
It's in the screen grab, here it is for testing:

`a storybook illustration of a (cat:dog:0.6), bear, picnic, random`

This is the output:

`a storybook illustration of a (cat`

**Comment by savisco** [comment]
Having the same issue.

`*\stable-diffusion-webui\extensions\stable-diffusion-webui-wildcards\wildcards\(masterpiece.txt not found for the __(masterpiece__ wildcard.`

![firefox_PXU1VnTxhM](https://user-images.githubusercontent.com/34283613/210186997-443c6352-0f6f-4142-bf2f-73745e01b992.png)

Prompt was simply
`(masterpiece:1.3), (best quality:1.1), (detailed:1.2), (realistic), absurdres, beautiful boat`

Is the latest issue I'm having when Dynamic Prompts is enabled. Seems like its looking for a wildcard for the very first word and cutting off the rest of the prompt.

**Comment by tusharbhutt** [comment]
> Is the latest issue I'm having when Dynamic Prompts is enabled. Seems like its looking for a wildcard for the very first word and cutting off the rest of the prompt.

I don't think looking at the first word as the wildcard, but stops after encountering the first colon

**Comment by adieyal** [comment]
X

> It's in the screen grab, here it is for testing:
> 
> `a storybook illustration of a (cat:dog:0.6), bear, picnic, random`
> 
> This is the output:
> 
> `a storybook illustration of a (cat`

is `(cat:dog:0.6)` meant to be prompt editing, i.e. `[cat:dog:0.6]`?

**Comment by adieyal** [comment]
> Prompt was simply `(masterpiece:1.3), (best quality:1.1), (detailed:1.2), (realistic), absurdres, beautiful boat`

That's bizarre - I tried it and it works fine - does it happen with other prompts as well?

**Comment by tusharbhutt** [comment]
> is `(cat:dog:0.6)` meant to be prompt editing, i.e. `[cat:dog:0.6]`?

Whoops, my mistake, you're correct.  Changing it fixed that error.  

However, this:

`a storybook illustration of a  (cat:1.3) and (dog:1.2), bear, picnic, random`

still gets output as

`a storybook illustration of a  (cat`

**Comment by tusharbhutt** [comment]
> That's bizarre - I tried it and it works fine - does it happen with other prompts as well?

I tried that prompt too, it does not work.  This is on the AUTO1111 commit 311354c0bb8930ea939d6aa6b3edd50c69301320

**Comment by adieyal** [comment]
I've disabled some of the parsing which I think was the cause of all the issues. Could you please upgrade to 1.5.15 and let me know if it resolves your issues?

**Comment by adieyal** [comment]
> I tried that prompt too, it does not work. This is on the AUTO1111 commit 311354c

Did you get the same wildcards error?

**Comment by Bonedragon1300** [comment]
ok that last update seems to have fixed the : issue on my end, but now it's not registering wildcards that are in parentheses or other wildcards
![Capture](https://user-images.githubusercontent.com/15267401/210188264-c863da12-f70d-4aff-91b1-a1bbe259ff09.PNG)

![Capture2](https://user-images.githubusercontent.com/15267401/210188272-b9f59979-f3e0-4eea-9990-94a3b1a9ad1e.PNG)

![Capture](https://user-images.githubusercontent.com/15267401/210188292-b96926b9-c612-4946-a778-1f6ac2a383a2.PNG)

**Comment by Bonedragon1300** [comment]
putting an extra space between the first ( and _, so instead of (__colors__), it's ( __colors__), seems to fix that, though I also notice that it's adding an extra space after the wildcard as well

**Comment by adieyal** [comment]
there is an outstanding issue where anything before the double underscores of the `__wildcard__` e.g. `,__` or `(__` is problematic - until its fixed the temporary workaround is to add a space like you did.

**Comment by Bonedragon1300** [comment]
ok, I'll have to keep an eye out for when that get's fixed, thanks for working on my problem so quickly

**Comment by tusharbhutt** [comment]
> > I tried that prompt too, it does not work. This is on the AUTO1111 commit 311354c
> 
> Did you get the same wildcards error?

The latest update is working with the workaround of putting a space before the double underscore of the wildcard filename and any parentheses, brackets, or braces. It seems like the error is only on the very first instance of parsing, so you don' need the space after the parentheses and the double underscore of any subsequent wildcard files in the prompt.

in other words:

`epic (sticker:1.2) of {__female astronaut__|__male settler__} , best, astonishing, impressive, outstanding, stunning, gorgeous, much detail, much wow, global dynamic lighting, cinematic, perfect, detailed textures, masterpiece`

works as well as:

`epic (sticker:1.2) of { __female astronaut__ | __male settler__ } , best, astonishing, impressive, outstanding, stunning, gorgeous, much detail, much wow, global dynamic lighting, cinematic, perfect, detailed textures, masterpiece`

However the output is slightly different because of those spaces

![image](https://user-images.githubusercontent.com/26264462/210198303-8092557d-bcfb-4030-9120-0daa7846df21.png)

**Comment by tusharbhutt** [comment]
> > > I tried that prompt too, it does not work. This is on the AUTO1111 commit 311354c
> > 
> > 
> > Did you get the same wildcards error?
> 
> It seems like the error is only on the very first instance of parsing, so you don't need the space after the parentheses and the double underscore of any subsequent wildcard files in the prompt.

I take that back,  It seems to work sometimes (maybe for braces?) but not other times.  Will need to test more

**Comment by adieyal** [comment]
I've tested the latest version against the examples in this thread. I think everything is now being parsed correctly. Could you please test and let me know?

**Comment by Puddz** [comment]
`(tilted head:1.0) [limited pallet:0.5] [hyper realistic::0.2] (matte color:1.0). atmospheric sparkling. close up.`

When Dynamic Prompts is turned on, the prompt only does the (tilted head:1.0) part of it, and cuts off everything else. 
Dynamic prompts turned off, and the prompt works as intended. 

Newest version of dynamic prompts and automatics webui.

**Comment by stysmmaker** [comment]
> (tilted head:1.0) [limited pallet:0.5] [hyper realistic::0.2] (matte color:1.0). atmospheric sparkling. close up.

This prompt works fine for me with Dynamic Prompts enabled on the latest version.

**Comment by Puddz** [comment]
Weird. Guess something is fucky with my installation then. Was working fine a few days ago and all I've done is update the installation and extension.

**Comment by Bonedragon1300** [comment]
needing spaces before a double underscore seems to be fixed, though it does still add extra spaces for some reason

**Comment by tusharbhutt** [comment]
> I've tested the latest version against the examples in this thread. I think everything is now being parsed correctly. Could you please test and let me know?


If I use spaces before and after the various tagging elements, the script works as expetcted.  I have updated everything (AUTO and Dynamic Prompts) in the last hour and I think an issue remains with this script and/or the Wildcards Manager extension _if there no spaces after the tagging elements_.  The extension uses or creates a folder:

`C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards`

and expects wildcards to be put in there.  It (or your script?) happily uses the wildcards in your script's folder;

`C:\Applications\StableDiffusion\extensions\sd-dynamic-prompts\wildcards`

**except when the prompt is within a tagging element, then _dynamic_prompting.py_ throws an error:**

For example, consider this prompt:

`epic (__mediumnew__) of (__hotpotmini__:1.2), professional (__genrenew__ genre:0.5), (__aestheticsnew__ style:0.5), __moodsnew__ mood, best, astonishing, impressive, outstanding, stunning, gorgeous, much detail, much wow, global dynamic lighting, cinematic, perfect, detailed textures, masterpiece`

The \_\_moodsnew\_\_ file works fine because it is not within parentheses, but for the rest I will get this error as AUTO tries to find each script if within a tagging element.

```
File C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\mediumnew.txt not found for the __mediumnew__ wildcard.
File C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\genrenew.txt not found for the __genrenew__ wildcard.
File C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\hotpotmini.txt not found for the __hotpotmini__ wildcard.
File C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\aestheticsnew.txt not found for the __aestheticsnew__ wildcard.
100%|██████████████████████████████████████████████████████████████████████████████████| 20/20 [00:21<00:00,  1.06s/it]
```

The interesting/frustrating thing is that placing any one of those wildcard files into the "C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\" folder does work, _but only if the wildcard file is not in a subfolder_.   

For example, in the "C:\Applications\StableDiffusion\extensions\sd-dynamic-prompts\wildcards" folder, I have a subfolder called "hotpot" and if I use the spacing workaround, there are no issues.  Copying the "hotpot" folder into C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\" and using propmts with no spaces does not work.  Placing the indivdual files from the "hotpot" folder into the "C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\" and prompting with not spaces _does work_.  The issue here is that it's hard to organize dozens of wildcard files without subfolders.

Turning the Wildcards manager extension on or off has no effect, and neither did creating a subfolder in the Collections folder in the "C:\Applications\StableDiffusion\extensions\sd-dynamic-prompts" folder

Apologies for the length of the response.  Here's a TL/DR of what works and does not

**Works:**

- If the wildcard is not within a tagging set (parentheses, braces, etc)
- If within a tag, using a space after the opening tag
- If within a tag, placing the wildcard file into the "C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\ folder"

**Does not work:**

- If within a tag, not using a space after the opening tag
- If within a tag, placing the wildcard file into a subfolder of the "C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\ folder" 
- Turning off the Wildcard Extension
- Creating a folder of wildcard files in the Collections Folder

**Comment by tusharbhutt** [comment]
Update: as of this morning, prompting seems to be working without the need for additional spacing in the prompt when using parentheses.  However, a prompt like this:

[fred|barney]

does not work (this the [alternating words feature ](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#alternating-words)format)

The feature works when the SD script is not enabled

**Comment by adieyal** [comment]
> needing spaces before a double underscore seems to be fixed, though it does still add extra spaces for some reason

@Bonedragon1300 Could you give me an example to test?

I've tried the following which all work:

```
(I love __colours__ roses)
(I love__colours__roses)
(I love__colours__:1.3) roses
I love {__colours__} roses
I love {__colours__|blue} roses
I love {|__colours__} roses
```

**Comment by adieyal** [comment]
@tusharbhutt thanks for the extensive testing. With regards your first message, the wildcards directory is by default stable-diffusion-webui/extensions/sd-dynamic-prompting/wildcards. It's possible to change this by setting a value in config.json but I don't think I've documented that so it's unlikely that you've changed it. You can find the exact path by clicking on the Wildcards Manager tab. Under bullet point 4, your directory is listed there.

Also, the collections folder is only used  to distribute the wildcards library. The extension doesn't look in that folder when processing the prompt. You need to manually copy the wildcards files into your wildcards directly or you can used the copy collection mechanism in the wildcards manager.

Regarding the alternating words bug `[fred|barney]` - thanks for picking this up. The reason it breaks is because the parser considers the `|` character to belong to a `{A|B|C}` construct and tries to process it. I'll look into fixing that.

**Comment by adieyal** [comment]
> **Does not work:**
> 
> 1. If within a tag, not using a space after the opening tag
> 2. If within a tag, placing the wildcard file into a subfolder of the "C:\Applications\StableDiffusion\extensions\stable-diffusion-webui-wildcards\wildcards\ folder"
> 3. Turning off the Wildcard Extension
> 4. Creating a folder of wildcard files in the Collections Folder

I just want to make sure whether these are still an issue.

1. I think this is working now
2. Mentioned above, that isn't the correct wildcards folder
3. I'm not sure what happens when you use both Dynamic Prompts and Wildcards at the same time. I imagine one of the extensions will step on the other's toes.
4. Discuss above - collections is not used for processing prompts

**Comment by tusharbhutt** [comment]
> I just want to make sure whether these are still an issue.
> 
>     1. I think this is working now
> 
>     2. Mentioned above, that isn't the correct wildcards folder

I'll have a look, but it does appear that not using a space after the tag is working now.  I'm not sure if that other folder is because of the Wildcards extension or if it's something from AUTO1111 itself.  The Wildcards manager extension uses your script (I think) so I'd hope they play nice together

**Comment by adieyal** [comment]
The Wildcards Manager is installed as part of Dynamic Prompts - but it looks for wildcards in extensions/sd-dynamic-prompts/wildcards which I think on your machine would be C:\Applications\StableDiffusion\extensions\sd-dynamic-prompts\wildcards

**Comment by tusharbhutt** [comment]
> @tusharbhutt I have fixed the alternating words bug, would you mind giving it a test? I haven't released it yet, to install it you need to run the following commands in your sd-dynamicprompts folder
> 
> ```
> git pull -a
> git checkout v2.0.2
> ```
> 
> If you want to return back to your current version, run this command
> 
> `git checkout main`
> 
> Let me know if it works for you.

Sure, will check after dinner, just got home from work.

**Comment by tusharbhutt** [comment]
Sorry, was sick.  I tried 2.02 and the program gave a very long error string after encountering a float16 error.  I'm not sure if it has anything todo with SD Dynamic Prompt or not.  I exited and not PyTorch cannot find CUDA so it errors out that a GPU does not exist and I cannot do anything.  Which is weird because I did not update any other component of AUTO1111.  I may have to reinstall AUTO111 if reverting Dynamic Prompt does not work
